linker/validate.go

```diff
@@ -76,3 +84,208 @@ func (r *result) validateExtension(fld protoreflect.FieldDescriptor, handler *re
 
 	return nil
 }
+
+func (r *result) validateJSONNamesInFile(handler *reporter.Handler) error {
+	for _, md := range r.FileDescriptorProto().GetMessageType() {
+		if err := r.validateJSONNamesInMessage(md, handler); err != nil {
+			return err
+		}
+	}
+	for _, ed := range r.FileDescriptorProto().GetEnumType() {
+		if err := r.validateJSONNamesInEnum(ed, handler); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *result) validateJSONNamesInMessage(md *descriptorpb.DescriptorProto, handler *reporter.Handler) error {
+	if err := r.validateFieldJSONNames(md, false, handler); err != nil {
+		return err
+	}
+	if err := r.validateFieldJSONNames(md, true, handler); err != nil {
+		return err
+	}
+
+	for _, nmd := range md.GetNestedType() {
+		if err := r.validateJSONNamesInMessage(nmd, handler); err != nil {
+			return err
+		}
+	}
+	for _, ed := range md.GetEnumType() {
+		if err := r.validateJSONNamesInEnum(ed, handler); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *result) validateJSONNamesInEnum(ed *descriptorpb.EnumDescriptorProto, handler *reporter.Handler) error {
```

Comment:
This looks to have nearly _all_ of https://github.com/jhump/protoreflect/pull/524.

Worth splitting this out into a separate PR (like the original patches)? If not, also add all of the other new tests from 524 to this PR?

Reply:
Went ahead and moved additional changes from https://github.com/jhump/protoreflect/pull/524 into this PR since they both were changing code in the same place. This should be ready for a re-review now.